### ISSUE TITLE
Associative Map i.e. dictionary

### ISSUE THREAD

**Issue by oxinabox** [google-research/dex-lang#332]
I was wanting a dictionary/set type 
So I thought I should make a tree based dictionary.
Nice traditional functional data structures, but can't do that without #331

A persistent hashmap is possible, particularly using effects.
I guess would need a Hashable typeclass, and then could use `rand` to whiten it.

**Comment by dougalm** [comment]
Right, I think the dexy thing to do would be to make a `Dict k v` type using an array-backed hash table, using effects to update it in-place when needed. You probably don't want the size of the table exposed in its type, so you'd do something like we do for `List` (or even just use `List` itself). Writing to lists in-place doesn't quite work yet, but @duvenaud has a really clever double-reference trick that should make it possible, pending a bugfix or two.

We'll eventually get to #331, but it would be more for things like an AST for an expression language -- data that's really not suited to being stored in arrays. We wouldn't expect to get great performance out of that sort of representation.
